### PR TITLE
Quad-precision math

### PR DESCRIPTION
**Pull request by dalg24**
Add support for GCC's 128-bit floating-point type

### CODE REVIEW

**Comment by mhoemmen** [comment]
I did this for Trilinos many years ago (but post adding Kokkos to Tpetra, I think?).

File: cmake/kokkos_arch.cmake [hunk 1]
```diff
@@ -17,6 +17,7 @@ ENDFUNCTION()
 KOKKOS_CFG_DEPENDS(ARCH COMPILER_ID)
 KOKKOS_CFG_DEPENDS(ARCH DEVICES)
 KOKKOS_CFG_DEPENDS(ARCH OPTIONS)
+#KOKKOS_CFG_DEPENDS(ARCH TPLS)
```
Comment (crtrott) [inline cmake/kokkos_arch.cmake:20]: What is this?

Comment (dalg24) [reply cmake/kokkos_arch.cmake:20]: I will drop.  That is the reason `KOKKOS_ENABLE_LIBQUADMATH` is not defined and I have to use `Kokkos_ENABLE_LIBQUADMATH`.  See the comment below.

File: .jenkins [hunk 2]
```diff
@@ -399,6 +399,8 @@ pipeline {
                                 -DKokkos_ENABLE_TESTS=ON \
                                 -DKokkos_ENABLE_OPENMP=ON \
                                 -DKokkos_ENABLE_LIBDL=OFF \
+                                -DKokkos_ENABLE_LIBQUADMATH=ON \
+                                -DCMAKE_PREFIX_PATH=/usr/local/lib/gcc/x86_64-unknown-linux-gnu/5.3.0 \
```
Comment (dalg24) [inline .jenkins:403]: This is a bit frustrating.  This is only needed for CMake, `g++` does not need to add anything to the include path to find `<quadmath.h>`.  Also `-lquadmath` on the link line would do.

File: core/src/impl/Kokkos_QuadPrecisionMath.hpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+#ifndef KOKKOS_QUAD_PRECISION_MATH_HPP
+#define KOKKOS_QUAD_PRECISION_MATH_HPP
+
+#include <Kokkos_NumericTraits.hpp>
+
+// NOTE probably belongs to <Kokkos_Macros.hpp>
+////////////////////////////////////////////////////////////////////////////////
+// Define if __float128 is supported on this host.
+#if defined(__FLOAT128__) || defined(__SIZEOF_FLOAT128__)
```
Comment (dalg24) [inline core/src/impl/Kokkos_QuadPrecisionMath.hpp:53]: I am tempted to just assert that this condition is true if `KOKKOS_ENABLE_LIBQUADMATH` is defined.
Any objection?

File: core/src/impl/Kokkos_QuadPrecisionMath.hpp [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+inline __float128 fdim(__float128 x, __float128 y) { return ::fdimq(x, y); }
+inline __float128 nanq(char const* arg) { return ::nanq(arg); }
+// Power functions
+inline __float128 pow(__float128 x, __float128 y) { return ::powq(x, y); }
+inline __float128 sqrt(__float128 x) { return ::sqrtq(x); }
+inline __float128 cbrt(__float128 x) { return ::cbrtq(x); }
+inline __float128 hypot(__float128 x, __float128 y) { return ::hypotq(x, y); }
+// Exponential functions
+inline __float128 exp(__float128 x) { return ::expq(x); }
+#if defined(KOKKOS_COMPILER_GNU) && (KOKKOS_COMPILER_GNU >= 910)
```
Comment (dalg24) [inline core/src/impl/Kokkos_QuadPrecisionMath.hpp:146]: Should be checking the stdlibc++ version instead

Comment (dalg24) [reply core/src/impl/Kokkos_QuadPrecisionMath.hpp:146]: Actually I am not sure...  There does not seem to be any macro to check the libquadmath version.

**Review by masterleinad — APPROVED** [review]
Looks OK to me (maybe apart from the `CMAKE_PREFIX_PATH` weirdness)

**Review by Rombur — APPROVED** [review]

**Review by crtrott — APPROVED** [review]

**Comment by ikalash** [comment]
> I did this for Trilinos many years ago (but post adding Kokkos to Tpetra, I think?).

Hi @mhoemmen !  hope you are doing well!  You may be happy to hear I am reviving all the float128 stuff you put in awhile ago, as well as making extensions for 'long double' scalar types.  Unfortunately there have been a number of issues getting the software stack to cooperate, sigh...

**Comment by ikalash** [comment]
Thanks for doing this @dalg24 !  Would Kokkos automatically detect if quadmath is enabled in Trilnos when being built as a part of Trilinos and set the configure flags appropriately?

**Comment by dalg24** [comment]
Yes
https://github.com/kokkos/kokkos/blob/cfb94d99fb6175547d989a79a7b3d0d5386ac566/cmake/kokkos_tpls.cmake#L70-L75

**Comment by ikalash** [comment]
Regarding adding the extended scalar types to Kokkos::rand - I understand the choice here.  The problem is that Kokkos::rand is used within basic Trilinos routines with the implied scalar type, e.g., https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp#L2345-L2350 .  I could try to change this if Trilnos folks are not opposed.  What is the best way to do this - do a downcase only in the case of a 'long double' or '__float128'?

**Comment by dalg24** [comment]
Use `Kokkos::Experimental::finite_max<T>::value` instead

**Comment by ikalash** [comment]
> Use `Kokkos::Experimental::finite_max<T>::value` instead

This is equivalent to Kokkos::rand?

**Comment by ikalash** [comment]
Ah ok, never mind, I see what you are suggesting.  Sorry, was looking in the wrong place.  I'll try that.

**Comment by ikalash** [comment]
So I get the following error when I switch to ```Kokkos::Experimental::finite_max<T>::value```:

```
/nightlyCDash/repos/Trilinos/packages/tpetra/core/src/Tpetra_MultiVector_def.hpp:2350:60: error: ‘value’ is not a member of ‘Kokkos::Experimental::finite_max<Kokkos::complex<double> >’
 2350 |     const IST max = Kokkos::Experimental::finite_max<IST>::value;
      |                                                            ^~~~~
```
I guess there are not complex implementations of finite_max?

**Comment by dalg24** [comment]
No but you can refer to the discussion on #4033 and do something like https://github.com/kokkos/kokkos/pull/4038/commits/9a78864d78d178251d34adf7827c9c4859af4397

**Comment by mhoemmen** [comment]
@ikalash wrote:

> hope you are doing well!

I am, thank you!  Hope you are too! : - )

> Unfortunately there have been a number of issues getting the software stack to cooperate, sigh...

Par for the course ; - )

> You may be happy to hear I am reviving all the `float128` stuff you put in awhile ago, as well as making extensions for `long double` scalar types. 

Excellent!  I do hope it works out!  These days I generally avoid `long double`, because on Windows it's the same as `double` (alas).  It could be interesting to work on a more portable solution, that defines a type alias promising 128-bit floating-point arithmetic.  This would align with the solution proposed by [P1467r4](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1467r4.html#alias-formats).

**Comment by ikalash** [comment]
> Excellent! I do hope it works out! These days I generally avoid `long double`, because on Windows it's the same as `double` (alas). It could be interesting to work on a more portable solution, that defines a type alias promising 128-bit floating-point arithmetic. This would align with the [P1467r4](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1467r4.html#alias-formats) C++ Standard Committee proposal.

Who uses Windows for scientific computing though??  (I know people do, I just always find it hard to believe).  I have been lamenting that you are not around to discuss what I've been doing!  It seems like there is a fair amount of interest in the extended scalar types within Trilinos, so it will be interesting to see if it goes anyway.

**Comment by mhoemmen** [comment]
@ikalash wrote:
> Who uses Windows for scientific computing though?? 

Ha! ; - P  Y'all need to remember that a lot of computation happens on the desktop.  Visual Studio's C++ support really took off in the past few years -- these days, we're using a lot of C++20 even without requiring the latest compiler version.

I've seen `long double` have 64 bits, 80 bits, or 128 bits, depending on the hardware and compiler.  It's not very nice if you're trying to write portable code.  One possible portable fall-back would be [QD](https://www.davidhbailey.com/dhbsoftware/).  Trilinos had QD support at one time, though it wasn't kept up.  These days, it might be better to look into [ReproBLAS](https://bebop.cs.berkeley.edu/reproblas/) techniques.  @jjellio knows more and has done GPU performance studies.

> I have been lamenting that you are not around to discuss what I've been doing! 

Likewise! : - )  (I'd be happy to talk more offline about contract work through my employer if you're interested; if not, no worries!)

> It seems like there is a fair amount of interest in the extended scalar types within Trilinos, so it will be interesting to see if it goes anyway.

Best wishes with that work!

**Pull request merged**